src/Raven.Studio/typescript/common/queryCompleter.ts

```diff
@@ -0,0 +1,279 @@
+        do {
+            const token = iterator.getCurrentToken();
+            if (!token) {
+                break;
+            }
+
+            switch (token.type) {
+                case "keyword":
+                    let keyword = token.value.toLowerCase();
+                    if (keyword === "desc" ||
+                        keyword === "asc" ||
+                        keyword === "and" ||
+                        keyword === "or")
+                        continue;
+
+                    return [keyword, operator, identifier, text, paren];
+                case "keyword.operator":
+                    operator = token.value;
+                    break;
+                case "identifier":
+                    identifier = token.value;
+                    break;
+                case "string":
+                    const indexName = token.value.substr(1, token.value.length - 2);
+                    identifier = indexName;
+                    break;
+                case "paren.lparen":
+                    paren++;
+                    break;
+                case "paren.rparen":
+                    paren--;
+                    break;
+                case "text":
+                    if (!identifier) {
+                        text = token.value;
+                    }
+                    break;
+            }
+        } while (iterator.stepBackward());
+
+        return [null, null, null, null, null];
+    }
+
+    complete(editor: AceAjax.Editor,
+             session: AceAjax.IEditSession,
+             pos: AceAjax.Position,
+             prefix: string,
+             callback: (errors: any[], worldlist: autoCompleteWordList[]) => void) {
+
+        const [lastKeyword, operator, identifier, text, paren] = this.getLastKeyword(session, pos);
+        if (!lastKeyword)
+            return;
+
+        switch (lastKeyword) {
+            case "from": {
+                if (identifier && text) {
+                    if (paren > 0) {
+                        // from (Collection, {show fields here})
+                        this.getIndexFields(session)
+                            .done((indexFields) => callback(null, indexFields.map(field => {
+                                return {name: field, value: field, score: this.defaultScore, meta: "field"};
+                            })));
+                        return;
+                    }
+
+                    return;
+                }
+
+                if (!prefix ||
+                    prefix.length === 0 ||
+                    prefix.startsWith("@")) {
+                    callback(null, [{name: "@all_docs", value: "@all_docs", score: this.defaultScore * 10, meta: "collection"}]);
+                    callback(null, [{name: "@system", value: "@system", score: this.defaultScore - 1, meta: "collection"}]);
+                }
+                callback(null, this.collectionsTracker.getCollectionNames().map(collection => {
+                    return {
```

Comment:
here you have example how to avoid functions which return list of [string, string, string, string, number, boolean]